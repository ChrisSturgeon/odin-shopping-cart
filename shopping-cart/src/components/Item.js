import '../styles/item.css';
import { useState } from 'react';

export default function Item(props) {
  const [qty, setQty] = useState(1);
  const [purchased, setPurchased] = useState(false);
  const path = props.artwork;

  const increment = () => {
    setQty((prevQty) => prevQty + 1);
  };

  const decrement = () => {
    if (qty >= 2) {
      setQty((prevQty) => prevQty - 1);
    }
  };

  return (
    <div className="item-card">
      <img src={path} alt="record artwork"></img>
      <div className="card-bottom">
        <div>{props.album}</div>
        <div>{props.artist}</div>
        {qty > 1 ? (
          <div>
            £{props.price} each / £
            {Math.round((props.price * qty + Number.EPSILON) * 100) / 100} total
          </div>
        ) : (
          <div>£{props.price}</div>
        )}

        <div className="card-qty">
          <button onClick={decrement}>-</button>
          <div>{qty}</div>
          <button onClick={increment}>+</button>
        </div>
        <button
          onClick={() => {
            props.add(props.id, qty);
            setQty(1);
            setPurchased(true);
          }}
        >
          {purchased ? 'Added!' : 'Add to cart'}
        </button>
      </div>
    </div>
  );
}
