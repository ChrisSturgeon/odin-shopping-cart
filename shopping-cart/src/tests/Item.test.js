import React from 'react';
import { render, screen } from '@testing-library/react';
import Item from '../components/Item';
import { BrowserRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';

describe('Item Tests', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(
      <Item price="17.99" album="Giant Steps" artist="John Coltrane" />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders album name', () => {
    render(<Item price="17.99" album="Giant Steps" artist="John Coltrane" />);
    expect(screen.getByText('Giant Steps')).toBeInTheDocument();
  });

  it('renders artist name', () => {
    render(<Item price="17.99" album="Giant Steps" artist="John Coltrane" />);
    expect(screen.getByText('John Coltrane')).toBeInTheDocument();
  });

  it('button renders correct text from initial state', () => {
    render(<Item price="17.99" album="Giant Steps" artist="John Coltrane" />);

    const addBtn = screen.getByRole('button', { name: 'Add to cart' });

    expect(addBtn).toHaveTextContent('Add to cart');
  });

  it('add to cart button renders correct text on click', () => {
    render(
      <Item
        price="17.99"
        album="Giant Steps"
        artist="John Coltrane"
        add={jest.fn()}
      />
    );
    const addBtn = screen.getByRole('button', { name: 'Add to cart' });

    userEvent.click(addBtn);

    expect(addBtn).toHaveTextContent('Success!');
  });

  it('quantity increases on increment button click', async () => {
    render(
      <Item
        price="17.99"
        album="Giant Steps"
        artist="John Coltrane"
        add={jest.fn()}
      />
    );
    const incrementBtn = screen.getByRole('button', { name: '+' });
    const quantity = screen.getByTestId('quantity');

    userEvent.click(incrementBtn);

    expect(quantity).toHaveTextContent('2');
  });

  it('quantity does not decrease below 1 on decrement button click', async () => {
    render(
      <Item
        price="17.99"
        album="Giant Steps"
        artist="John Coltrane"
        add={jest.fn()}
      />
    );
    const decrementBtn = screen.getByRole('button', { name: '-' });
    const quantity = screen.getByTestId('quantity');

    userEvent.click(decrementBtn);

    expect(quantity).toHaveTextContent('1');
  });

  it('decrement btn decreases quantity', async () => {
    render(
      <Item
        price="17.99"
        album="Giant Steps"
        artist="John Coltrane"
        add={jest.fn()}
      />
    );
    const incrementBtn = screen.getByRole('button', { name: '+' });
    const decrementBtn = screen.getByRole('button', { name: '-' });
    const quantity = screen.getByTestId('quantity');

    userEvent.click(incrementBtn);
    userEvent.click(incrementBtn);
    userEvent.click(decrementBtn);

    expect(quantity).toHaveTextContent('2');
  });
});
