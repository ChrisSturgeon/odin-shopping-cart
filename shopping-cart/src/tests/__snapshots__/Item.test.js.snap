// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item Tests matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="item-card"
  >
    <img
      alt="record artwork"
    />
    <div
      class="card-bottom"
    >
      <div>
        Giant Steps
      </div>
      <div
        class="artist"
      >
        John Coltrane
      </div>
      <div>
        Â£17.99
      </div>
      <div
        class="card-qty"
      >
        <button>
          -
        </button>
        <div
          data-testid="quantity"
        >
          1
        </div>
        <button>
          +
        </button>
      </div>
      <button>
        Add to cart
      </button>
    </div>
  </div>
</DocumentFragment>
`;
